### PR TITLE
AAA handler session state stats

### PR DESCRIPTION
**Pull request by fholzhauser**

### CODE REVIEW

**Comment by fholzhauser** [comment]
Thanks Bengt ! Indeed I was considering that, but I did not want to go too far beyond the scope of this feature with the changes, so at the end I didn't. However your comment reminded me to add tests for radius handler too.

File: rebar.config [hunk 1]
```diff
@@ -4,6 +4,7 @@
 {deps, [
 	{setup, "2.0.2"},
 	{jobs, "0.9.0"},
+	{prometheus, "4.4.1"},
```
Comment (RoadRunnr) [inline rebar.config:7]: ergw has 4.6.0

File: doc/aaa_session_metrics.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+AAA diameter session metrics
+================================
+
+ergw_aaa exposes prometheus gauge metric for the states of the different diameter handlers
```
Comment (RoadRunnr) [inline doc/aaa_session_metrics.md:4]: there is a trailing whitespace in that line (says git diff)

File: src/ergw_aaa_app.erl [hunk 3]
```diff
@@ -20,6 +20,7 @@ start(_StartType, _StartArgs) ->
     case ergw_aaa_sup:start_link() of
 	{ok, _} = Ret ->
 	    Config = ergw_aaa_config:load_config(),
+		prometheus_declare(),
```
Comment (RoadRunnr) [inline src/ergw_aaa_app.erl:23]: indenetion is wrong

File: src/ergw_aaa_gx.erl [hunk 4]
```diff
@@ -616,3 +618,6 @@ make_CCR(Type, Session, Opts, #state{request_number = ReqNumber}) ->
 		   'CC-Request-Number'   => ReqNumber},
     Avps = from_session(Session, Avps1),
     ['CCR' | Avps ].
+
+get_state_atom(#state{state = State}) ->
+    State.
```
Comment (RoadRunnr) [inline src/ergw_aaa_gx.erl:623]: no newline at end of file

File: src/ergw_aaa_radius.erl [hunk 5]
```diff
@@ -755,3 +756,6 @@ send_request(Req, Session, #{server := NAS, retries := Retries, timeout := Timeo
 			{retries, Retries},
 			{timeout, Timeout}],
     eradius_client:send_request(NAS, Req, RadiusClientOpts).
+
+get_state_atom(#{'State' := State}) ->
+    State.
```
Comment (RoadRunnr) [inline src/ergw_aaa_radius.erl:761]: newline

File: src/ergw_aaa_rf.erl [hunk 6]
```diff
@@ -656,3 +657,6 @@ create_ACR(Type, Session, #{now := Now} = Opts, #state{record_number = RecNumber
     {Avps4, State2} = traffic_data(Avps3, State1),
     {Avps, State} = ran_secondary_rat_usage_report(Avps4, State2),
     {['ACR' | from_session(Session, Avps)], State#state{record_number = RecNumber + 1}}.
+
+get_state_atom(#state{state = State}) ->
+    State.
```
Comment (RoadRunnr) [inline src/ergw_aaa_rf.erl:662]: newline

File: src/ergw_aaa_gx.erl [hunk 7]
```diff
@@ -277,23 +279,23 @@ handle_cca({error, no_connection}, Session, Events,
     Avps = maps:get(Answer, Answers, #{'Result-Code' =>
 					   [?'DIAMETER_BASE_RESULT-CODE_AUTHORIZATION_REJECTED']}),
     handle_cca(['CCA' | Avps], Session, Events, Opts, State#state{state = peer_down});
-handle_cca({error, no_connection}, Session, Events,
+handle_cca({error, timeout}, Session, Events,
```
Comment (RoadRunnr) [inline src/ergw_aaa_gx.erl:282]: Is that change correct?

File: src/ergw_aaa_nasreq.erl [hunk 8]
```diff
@@ -532,3 +532,6 @@ create_STR(Session, Opts) ->
 		   'Auth-Application-Id' => ?'DIAMETER_APP_ID_NASREQ'},
     Avps = from_session(Session, Avps1),
     ['STR' | Avps].
+
+get_state_atom(#state{state = State}) ->
+    State.
```
Comment (RoadRunnr) [inline src/ergw_aaa_nasreq.erl:537]: newline

File: src/ergw_aaa_ro.erl [hunk 9]
```diff
@@ -789,3 +790,6 @@ make_CCR(Type, Session, #{now := Now} = Opts, #state{request_number = ReqNumber}
 	   end,
     Avps = Avps2#{'Multiple-Services-Credit-Control' => maps:values(MSCC)},
     ['CCR' | Avps ].
+
+get_state_atom(#state{state = State}) ->
+    State.
```
Comment (RoadRunnr) [inline src/ergw_aaa_ro.erl:795]: newline

File: src/ergw_aaa_static.erl [hunk 10]
```diff
@@ -88,3 +90,6 @@ handle_response(_Procedure, #{'Result-Code' := Code}, Session, Events, State) ->
 handle_response(_Procedure, Response, Session, Events, State) ->
     ?LOG(error, "Response: ~p", [Response]),
     {Response, Session, Events, State}.
+
+get_state_atom(_) ->
+    stopped.
```
Comment (RoadRunnr) [inline src/ergw_aaa_static.erl:95]: newline

File: src/ergw_aaa_session.erl [hunk 11]
```diff
@@ -438,5 +438,30 @@ step({Service, SvcOpts}, Procedure, #data{handlers = HandlersS,
...
+	{_, stopped} ->
+	    prometheus_gauge:dec(aaa_sessions_total, [Handler, CS]);
+    {undefined, _} ->
+        prometheus_gauge:inc(aaa_sessions_total, [Handler, NS]);
+    {stopped, _} ->
+        prometheus_gauge:inc(aaa_sessions_total, [Handler, NS]);
+	{_, _} ->
+	    prometheus_gauge:dec(aaa_sessions_total, [Handler, CS]),
+	    prometheus_gauge:inc(aaa_sessions_total, [Handler, NS])
+    end.
```
Comment (RoadRunnr) [inline src/ergw_aaa_session.erl:462]: * indention is strange
* I would write it like this (untested):

```suggestion
aaa_state_stats_dec(_, From)
  when From =:= undefined; From =:= stopped ->
    ok;
aaa_state_stats_dec(Handler, From) ->
    prometheus_gauge:dec(aaa_sessions_total, [Handler, From]).

aaa_state_stats_inc(_, To)
  when To =:= undefined; To =:= stopped ->
    ok;
aaa_state_stats_inc(Handler, To) ->
    prometheus_gauge:inc(aaa_sessions_total, [Handler, To]).

aaa_state_stats(Handler, CurrentState, NewState) ->
    From = get_handler_state(Handler, CurrentState),
    To = get_handler_state(Handler, NewState),
    if From /= To ->
	    aaa_state_stats_dec(Handler, From),
	    aaa_state_stats_inc(Handler, To);
       From == To ->
	    ok
    end.
```

File: test/diameter_Gx_SUITE.erl [hunk 12]
```diff
@@ -115,16 +115,18 @@ init_per_suite(Config0) ->
 
 end_per_suite(Config) ->
     meck_unload(Config),
+	application:stop(prometheus),
```
Comment (RoadRunnr) [inline test/diameter_Gx_SUITE.erl:118]: indention

File: test/diameter_Gx_SUITE.erl [hunk 13]
```diff
@@ -197,15 +199,21 @@ simple_session(Config) ->
     Stats0 = get_stats(?SERVICE),
 
     {ok, SId} = ergw_aaa_session_sup:new_session(self(), Session),
+	?equal([], prometheus_gauge:values(default, aaa_sessions_total)),
```
Comment (RoadRunnr) [inline test/diameter_Gx_SUITE.erl:202]: indention

File: test/diameter_Gy_SUITE.erl [hunk 14]
```diff
@@ -553,46 +589,6 @@ rate_limit(_Config) ->
...
-    ?equal(Ok, lists:sum(ReceivedRate)),
-
-    %% all 50 requests were either rate limited or ok (no other error)
-    ?equal(50, Ok + RateLimited),
-
-    ?match(0, outstanding_reqs()),
-    meck_validate(Config),
-    ?CLEAR_TC_INFO(),
-    ok.
-
```
Comment (RoadRunnr) [inline test/diameter_Gy_SUITE.erl:595]: why do you remove this?

File: test/ergw_aaa_test_lib.erl [hunk 15]
```diff
@@ -160,3 +161,15 @@ diff_stats(S1, S2) ->
...
+get_session_stats() ->
+    lists:sort([{Handler, State, Value} || 
+     {[{"handler", Handler}, {"state", State}], Value} <- 
+     prometheus_gauge:values(default, aaa_sessions_total)]).
+
+reset_session_stats() ->
+    %% there seems to be no easier way doing this ...
+    [prometheus_gauge:remove(default, aaa_sessions_total, [Handler, State]) ||
+     {[{"handler", Handler}, {"state", State}], _} <- 
+     prometheus_gauge:values(default, aaa_sessions_total)].
```
Comment (RoadRunnr) [inline test/ergw_aaa_test_lib.erl:175]: trailing whitespace

File: src/ergw_aaa_gx.erl (hunk 7 above)
Comment (fholzhauser) [reply src/ergw_aaa_gx.erl:282]: Yes, it is a fixing a wrong function clause match, however certainly has nothing to do with this feature. Shall I "unfix" it and make a new PR for clarity?

File: test/diameter_Gy_SUITE.erl (hunk 14 above)
Comment (fholzhauser) [reply test/diameter_Gy_SUITE.erl:595]: This test is not used, as the current rate limiter does not distinguish request types and the other (used) rate limiting test is functionally very similar to this one. However this change is not related to the scope of this PR, so I'll add this back here and open a separate PR for it.

File: rebar.config (hunk 1 above)
Comment (fholzhauser) [reply rebar.config]: updated

File: doc/aaa_session_metrics.md (hunk 2 above)
Comment (fholzhauser) [reply doc/aaa_session_metrics.md]: fixed

**Pull request merged**